### ISSUE TITLE
Bug Report: weight_string() does not add the table identifier in a scattered join/orderby query

### ISSUE THREAD

**Issue by tanjinx** [vitessio/vitess#14067]
### Overview of the Issue

In a scattered join/order-by query where the two tables joined have a column with the same name, the weight_string() does not add the table identifier for that column, which resulted in an ambiguous error:

mysql> select A.created, B.created as toplevel_created from A, B where B.id=A.user_id and A.created>3675172300 and B.created>3675172300 order by created asc limit 10;

ERROR 1052 (23000): target: keyspace1.-10.primary: vttablet: rpc error: code = InvalidArgument desc = Column 'created' in field list is ambiguous (errno 1052) (sqlstate 23000) (CallerID: vt_appdebug_a): Sql: "select A.created, B.created as toplevel_created, weight_string(created) from A, B where B.created > :vtg1 and A.created > :vtg1 and B.id = A.user_id order by created asc limit :__upper_limit", BindVars: {__upper_limit: "type:INT64 value:\"10\""vtg1: "type:INT64 value

v12 :
```
mysql god@localhost:keyspace1> EXPLAIN FORMAT=vitess select A.created, B.created as toplevel_created from A, B where B.id=A.user_id and A.created>3675172300 and B.created>3675172300 order by created asc limit 10;
+----------------+-------------------+------------+---------------+--------------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
| operator       | variant           | keyspace   | destination   | tabletType   | query                                                                                                                                                      |
|----------------+-------------------+------------+---------------+--------------+------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Limit          |                   |            |               | UNKNOWN      |                                                                                                                                                            |
| └─ Sort        | Memory            |            |               | UNKNOWN      |                                                                                                                                                            |
|    └─ Join     | Join              |            |               | UNKNOWN      |                                                                                                                                                            |
|       ├─ Route | SelectScatter     | keyspace1  |               | UNKNOWN      | select B.created as toplevel_created, B.id from B where B.created > 3675172300                                                             |
|       └─ Route | SelectEqualUnique | keyspace   |               | UNKNOWN      | select A.created, weight_string(A.created) from A where A.user_id = :users_id and A.created > 3675172300 |
+----------------+-------------------+------------+---------------+--------------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
5 rows in set
Time: 0.003s
```
v14: (14.0.5)
```
mysql> select A.created, B.created as toplevel_created from A, B where B.id=A.user_id and A.created>3675172300 and B.created>3675172300 order by created asc limit 10;
ERROR 1052 (23000): target: keyspace1.-10.primary: vttablet: rpc error: code = InvalidArgument desc = Column 'created' in field list is ambiguous (errno 1052) (sqlstate 23000) (CallerID: vt_appdebug_a): Sql: "select A.created, B.created as toplevel_created, weight_string(created) from A, B where B.created > :vtg1 and A.created > :vtg1 and B.id = A.user_id order by created asc limit :__upper_limit", BindVars: {__upper_limit: "type:INT64 value:\"10\""vtg1: "type:INT64 value
mysql god@localhost:keyspace1> EXPLAIN FORMAT=vitess select A.created, B.created as toplevel_created from
                           -> A, B where B.id=A.user_id and A.created>3675172300 and
                           -> B.created>3675172300 order by created asc limit 10;
+------------+-----------+------------+---------------+--------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| operator   | variant   | keyspace   | destination   | tabletType   | query                                                                                                                                                                                                                                                            |
|------------+-----------+------------+---------------+--------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Limit      |           |            |               | UNKNOWN      |                                                                                                                                                                                                                                                                  |
| └─ Route   | Scatter   | keyspace1   |               | UNKNOWN      | select A.created, B.created as toplevel_created, weight_string(created) from A, B where B.created > 3675172300 and A.created > 3675172300 and B.id = A.user_id order by created asc limit :__upper_limit |
+------------+-----------+------------+---------------+--------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set
Time: 0.002s
```


### Reproduction Steps

create table t1 (id int, created timestamp);
create table t2 (user_id int, created timestamp);

select t1.created, t2.created as top_created from t1, t2 where t1.id = t2.user_id order by created; // a scatter query

### Binary Version

```sh
14.0.5
```


### Operating System and Environment details

```sh
linux amd64
```


### Log Fragments

_No response_

**Comment by henryr** [comment]
Just to add some more observations:

1. This doesn't seem to be fixed in any version up to v17.0.2. 
2. If we use an explicit column alias for the ordered-by column, the `WEIGHT_STRING` argument is correctly rewritten:

```
select A.id as id, B.id as foo from user A, user B WHERE A.id = B.id AND A.id IN (1,2,3) order by id asc limit 10;
```

becomes:

```
select A.id as id, B.id as foo, weight_string(A.id) from `user` as A, `user` as B where A.id in (1, 2) and A.id = B.id order by id asc limit 10
... etc
```

so the error seems to be that `SELECT A.id ... order by id` is not analyzed as a column alias expression, but instead a column _name_ expression.

**Comment by systay** [comment]
Hi @henryr & @tanjinx 

We've done a quite invasive update of the planner in the V18 release time frame. I just tried this on `main`, and it produces a good plan:

```json
{
  "QueryType": "SELECT",
  "Original": "select u.foo, ue.foo as sumfin from user u, user_extra ue where u.A = ue.A order by foo",
  "Instructions": {
    "OperatorType": "Join",
    "Variant": "Join",
    "JoinColumnIndexes": "L:0,R:0",
    "JoinVars": {
      "u_A": 1
    },
    "TableName": "`user`_user_extra",
    "Inputs": [
      {
        "OperatorType": "Route",
        "Variant": "Scatter",
        "Keyspace": {
          "Name": "user",
          "Sharded": true
        },
        "FieldQuery": "select u.foo, u.A, weight_string(u.foo) from `user` as u where 1 != 1",
        "OrderBy": "(0|2) ASC",
        "Query": "select u.foo, u.A, weight_string(u.foo) from `user` as u order by foo asc",
        "Table": "`user`"
      },
      {
        "OperatorType": "Route",
        "Variant": "Scatter",
        "Keyspace": {
          "Name": "user",
          "Sharded": true
        },
        "FieldQuery": "select ue.foo as sumfin from user_extra as ue where 1 != 1",
        "Query": "select ue.foo as sumfin from user_extra as ue where ue.A = :u_A",
        "Table": "user_extra"
      }
    ]
  },
  "TablesUsed": [
    "user.user",
    "user.user_extra"
  ]
}
```

The query works well in the next version of Vitess, version 18. (https://github.com/vitessio/vitess/releases/tag/v18.0.0-rc1 is the RC of this release)

Until you switch to V18, you can fix these queries by using the V3 planner.

I hope this is an acceptable solution for you.

**Comment by GuptaManan100** [comment]
@systay Will do a separate fix for this in release-17.0
